### PR TITLE
Support unmanaged SFDX dependencies

### PR DESCRIPTION
**Pull request by davidmreed**
# Critical Changes

# Changes

- CumulusCI now supports deploying unmanaged dependencies in SFDX source format.
- The `create_package_version` task now handles dependencies that use a `zip_url`.

# Issues Closed

### CODE REVIEW

File: cumulusci/core/sfdx.py [hunk 1]
```diff
@@ -89,3 +97,57 @@ def get_default_devhub_username():
...
+
+@contextlib.contextmanager
+def convert_sfdx_source(
+    path: T.Optional[PathLike], name: T.Optional[str], logger: logging.Logger
+):
+    mdapi_path = None
+    with contextlib.ExitStack() as stack:
+        # Convert SFDX -> MDAPI format if path exists but does not have package.xml
+        if (
+            len(os.listdir(path or "."))
```
Comment (davidmreed) [inline cumulusci/core/sfdx.py:135]: To be tested: does this work on WIndows?

File: cumulusci/core/dependencies/dependencies.py [hunk 2]
```diff
@@ -523,12 +530,65 @@ def install(self, context: BaseProjectConfig, org: OrgConfig):
...
+            not self.subfolder
+            and get_source_format_for_zipfile(zip_src, "src") is SourceFormat.MDAPI
+        ):
+            zip_extract_subfolder = "src"
+        elif (
+            self.subfolder
+            and get_source_format_for_zipfile(zip_src, self.subfolder)
+            is SourceFormat.MDAPI
+        ):
+            zip_extract_subfolder = self.subfolder
```
Comment (jstvz) [inline cumulusci/core/dependencies/dependencies.py:550]: Could this be simplified with Extract Variable? For example:
```suggestion
        zip_extract_subfolder = self.subfolder or "src"
        source_format = get_source_format_for_zipfile(zip_src, zip_extract_subfolder)
        if source_format is not SourceFormat.MDAPI:
            zip_extract_subfolder = None
```

**Review by jstvz — COMMENTED** [review]
Hopefully these all made sense. Still working on getting the Windows AWS WorkSpace set up correctly.

---
### Test cases

| # | scenario             | pass |
|---|----------------------|------|
| 0 | project pre/post     | ✅   |
| 1 | dependency pre/post  | ❎   |
| 2 | dependency unmanaged | ✅   |
| 3 | dependency subfolder | ✅   |
| 4 | zip_url              | ❎   |


<details>
<summary>Case 0</summary>

Grants-management v1.8 had a source-formatted directory under
`unpackaged/post/enhanced_application`.

```console
$ git checkout release/1.8
HEAD is now at 29a86e39 Merge pull request #555 from SalesforceFoundation/feature/230__translations
$ cci flow run dev_org --org dev
[...]
Completed flow 'dev_org' on org dev successfully!
```

</details>
<details>
<summary>Case 1</summary>

From PMM `main`:

```diff
diff --git a/cumulusci.yml b/cumulusci.yml
index a49b166e..0e7dbf40 100644
--- a/cumulusci.yml
+++ b/cumulusci.yml
@@ -15,6 +15,8 @@ project:
     dependencies:
         - namespace: sfdobase
           version: 1.0
+        - github: https://github.com/SalesforceFoundation/Grants-Management
+          tag: release/1.8
     dependency_resolutions:
         preproduction: include_beta

diff --git a/orgs/dev_exp.json b/orgs/dev_exp.json
index 5bd68818..f2a2d29d 100644
--- a/orgs/dev_exp.json
+++ b/orgs/dev_exp.json
@@ -1,7 +1,7 @@
 {
     "orgName": "Program Management Module - Experience Cloud Dev Org",
     "edition": "Developer",
-    "features": ["Communities"],
+    "features": ["GRANTSMANAGEMENTUSER", "Communities", "PersonAccounts", "DebugApex"],
     "settings": {
         "lightningExperienceSettings": {
             "enableS1DesktopEnabled": true
```

Running `cci flow run dependencies --org dev_exp` throws the following
exception:

```python-traceback
Traceback (most recent call last):
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/salesforce_api/metadata.py", line 68, in __call__
    return self._process_response(response)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/salesforce_api/metadata.py", line 514, in _process_response
    raise MetadataComponentFailure(log, response)
cumulusci.salesforce_api.exceptions.MetadataComponentFailure: Update of None package.xml: Error: No package.xml found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/cli/cci.py", line 78, in main
    cli(args[1:], standalone_mode=False, obj=runtime)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/cli/runtime.py", line 176, in new_func
    func(runtime, *args, **kw)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/cli/flow.py", line 158, in flow_run
    coordinator.run(org_config)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/core/flowrunner.py", line 422, in run
    self._run_step(step)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/core/flowrunner.py", line 463, in _run_step
    raise result.exception  # PY3: raise an exception type we control *from* this exception instead?
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/core/flowrunner.py", line 250, in run_step
    task()
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/core/tasks.py", line 149, in __call__
    self.result = self._run_task()
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/tasks/salesforce/update_dependencies.py", line 199, in _run_task
    self._install_dependency(d)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/tasks/salesforce/update_dependencies.py", line 211, in _install_dependency
    dependency.install(self.project_config, self.org_config)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/core/dependencies/dependencies.py", line 532, in install
    return api()
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/salesforce_api/metadata.py", line 70, in __call__
    raise MetadataParseError(
cumulusci.salesforce_api.exceptions.MetadataParseError: Could not process MDAPI response: Update of None package.xml: Error: No package.xml found
```

</details>
<details>
<summary>Case 2</summary>

From K-12:

```diff
diff --git a/cumulusci.yml b/cumulusci.yml
index e1cac28..4cf591d 100644
--- a/cumulusci.yml
+++ b/cumulusci.yml
@@ -14,6 +14,7 @@ project:
         repo_url: https://github.com/SalesforceFoundation/k12-architecture-kit
     dependencies:
         - github: "https://github.com/SalesforceFoundation/EDA"
+          unmanaged: True
     dependency_resolutions:
         #preproduction: include_beta
         preproduction: commit_status
```

Running `cci flow run dependencies --org dev` deploys `force-app`
successfully, but EDA's post-install gets tripped up on
`unpackaged/post` and throws the following exception:

```python-traceback
Traceback (most recent call last):
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/salesforce_api/metadata.py", line 68, in __call__
    return self._process_response(response)
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/salesforce_api/metadata.py", line 514, in _process_response
    raise MetadataComponentFailure(log, response)
cumulusci.salesforce_api.exceptions.MetadataComponentFailure: Update of RecordType Case.Incident: Error on line 21, col 25: Global value set '%%%NAMESPACE%%%Category__c' in record type 'Incident' on the entity 'Case' cannot be resolved

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/cli/cci.py", line 78, in main
    cli(args[1:], standalone_mode=False, obj=runtime)
  File "/Users/jestevez/.pyenv/versions/3.9.5/envs/cci-d/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jestevez/.pyenv/versions/3.9.5/envs/cci-d/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/jestevez/.pyenv/versions/3.9.5/envs/cci-d/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jestevez/.pyenv/versions/3.9.5/envs/cci-d/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jestevez/.pyenv/versions/3.9.5/envs/cci-d/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jestevez/.pyenv/versions/3.9.5/envs/cci-d/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/jestevez/.pyenv/versions/3.9.5/envs/cci-d/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/cli/runtime.py", line 176, in new_func
    func(runtime, *args, **kw)
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/cli/flow.py", line 158, in flow_run
    coordinator.run(org_config)
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/core/flowrunner.py", line 422, in run
    self._run_step(step)
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/core/flowrunner.py", line 463, in _run_step
    raise result.exception  # PY3: raise an exception type we control *from* this exception instead?
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/core/flowrunner.py", line 250, in run_step
    task()
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/core/tasks.py", line 149, in __call__
    self.result = self._run_task()
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/tasks/salesforce/update_dependencies.py", line 199, in _run_task
    self._install_dependency(d)
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/tasks/salesforce/update_dependencies.py", line 211, in _install_dependency
    dependency.install(self.project_config, self.org_config)
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/core/dependencies/dependencies.py", line 592, in install
    return api()
  File "/Users/jestevez/work/rel/CumulusCI/cumulusci/salesforce_api/metadata.py", line 70, in __call__
    raise MetadataParseError(
cumulusci.salesforce_api.exceptions.MetadataParseError: Could not process MDAPI response: Update of RecordType Case.Incident: Error on line 21, col 25: Global value set '%%%NAMESPACE%%%Category__c' in record type 'Incident' on the entity 'Case' cannot be resolved
```

</details>
<details>
<summary>Case 3</summary>

From K-12:

```diff
diff --git a/cumulusci.yml b/cumulusci.yml
index e1cac28..8896c22 100644
--- a/cumulusci.yml
+++ b/cumulusci.yml
@@ -14,6 +14,7 @@ project:
         repo_url: https://github.com/SalesforceFoundation/k12-architecture-kit
     dependencies:
         - github: "https://github.com/SalesforceFoundation/EDA"
+          subfolder: force-app/main/settingsComponentLibrary
     dependency_resolutions:
         #preproduction: include_beta
         preproduction: commit_status
```

Running `cci flow run dependencies --org dev` succeeds.

</details>
<details>
<summary>Case 4</summary>

This one's probably expected. From K-12:

```diff
diff --git a/cumulusci.yml b/cumulusci.yml
index e1cac28..6380ae4 100644
--- a/cumulusci.yml
+++ b/cumulusci.yml
@@ -13,7 +13,7 @@ project:
         prefix_release: release/
         repo_url: https://github.com/SalesforceFoundation/k12-architecture-kit
     dependencies:
-        - github: "https://github.com/SalesforceFoundation/EDA"
+      - zip_url: https://github.com/SalesforceFoundation/EDA/archive/refs/tags/rel/1.116.zip
     dependency_resolutions:
         #preproduction: include_beta
         preproduction: commit_status
```

Running `cci flow run dependencies --org dev` succeeds.
```python-traceback
Traceback (most recent call last):
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/salesforce_api/metadata.py", line 68, in __call__
    return self._process_response(response)
  File "/usr/local/Cellar/cumulusci/3.39.0/libexec/lib/python3.9/site-packages/cumulusci/salesforce_api/metadata.py", line 514, in _process_response
    raise MetadataComponentFailure(log, response)
cumulusci.salesforce_api.exceptions.MetadataComponentFailure: Update of None package.xml: Error: No package.xml found
```
</details>

**Comment by davidmreed** [comment]
@jstvz Thanks for exhaustive testing of this one. 

Case 4 is, as you surmised, expected. GitHub zipballs have an extra layer of structure that the `zip_url` dependency installer does not understand. I did test it with a gimmick to remove the extra folder structure and it worked.

I am unable to reproduce the failure in Case 1. Can you help me figure out the proximate cause? I was surprised to see that failure because that code path is tested pretty heavily.

**Comment by jstvz** [comment]
> @jstvz Thanks for exhaustive testing of this one.
> 
> Case 4 is, as you surmised, expected. GitHub zipballs have an extra layer of structure that the `zip_url` dependency installer does not understand. I did test it with a gimmick to remove the extra folder structure and it worked.
> 
> I am unable to reproduce the failure in Case 1. Can you help me figure out the proximate cause? I was surprised to see that failure because that code path is tested pretty heavily.

The traceback gives us a clue `"/usr/local/Cellar/cumulusci/3.39.0` 🤦🏾 

Wait one.

**Review by jstvz — APPROVED** [review]
Case 1 works. Case 4 still fails, but with a SFDX error instead of MDAPI. I updated the original review. Everything else looks good.

**Pull request merged**